src/snmalloc/pal/pal_windows.h

```diff
@@ -27,6 +27,263 @@
 #    endif
 #  endif
 
+/**
+ * In order to properly cleanup our pagemap reservation,
+ * we need to make sure we do it is done after all other
+ * allocations have been freed.
+ *
+ * One way to guarantee that the reservations get released
+ * at the absolute end of the program is to use atexit.
+ * The atexit() takes function pointers which are called
+ * during exit in LIFO order. To make sure that we are
+ * the first call in the atexit chain, we use the pragma
+ * init_seg(".CRT$XCB")  which ensures that statics in
+ * this translation unit get initialized first and that
+ * atexit() in the constructor gets called first. Then,
+ * on program or DLL exit, we know that our function will
+ * be called last.
+ */
+#  pragma warning(disable : 4075)
+#  pragma init_seg(".CRT$XCB")
+
+// Simple max function (avoiding <algorithm>)
+static size_t max_size_t(size_t a, size_t b)
+{
+  return a > b ? a : b;
+}
+
+/**
+ * This VirtualVector class is an implementation of
+ * a vector, but does not use new/malloc or any other
+ * STL containers. These cannot be used in the
+ * function given to atexit() when the pragma
+ * init_seg(".CRT$XCB") is used, because the CRT
+ * might not be fully initialized yet. The segments
+ * compiler, lib, user cannot be used because of
+ * CRT internals like std::locale facets
+ */
+template<typename T>
+class VirtualVector
+{
+  T* data = nullptr;
+  size_t size = 0;
+  size_t committed_elements = 0;
+  size_t reserved_elements = 0;
+
+public:
+  VirtualVector(size_t reserve_elems = 64, size_t initial_commit = 32)
```

Comment:
Should the initial sizes be such that it is a Page?  
  (PAGE_SIZE / sizeof(T))
Perhaps, I guess there are issues if sizeof(T) is larger than a page. So would probably want a max in there to stop it becoming 0.